domain_utils/domain_utils.py

```diff
@@ -48,88 +30,181 @@ def is_ip_address(hostname):
+    """
+    Returns a url stripped to just the beginning and end, or more formally,
+    ``(scheme)?+(netloc|hostname)+(path)?``.
+    For example ``https://my.domain.net/a/path/to/a/file.html#anchor?a=1``
+    becomes ``my.domain.net/a/path/to/a/file.html``
+    URL parsing is done using std lib
+    `urllib.parse.urlparse
+    <https://docs.python.org/3.8/library/urllib.parse.html>`_.
+    Empty scheme e.g. ``my.domain.cloudfront.net``
+    are assumed to be http schemes.
+    If a URL has a port but no scheme, urlparse determines the scheme to
+    be the hostname and we do not handle this special case. In this case,
+    the url will be treated as a non-http scheme and the return value will
+    be determined by the ``drop_non_http`` setting.
+    :param url: URL to be parsed
+    :type url: str
+    :param scheme: If ``True``, scheme will be prepended in
+        returned result. Defaults is ``False``.
+    :type scheme: bool, optional
+    :param drop_non_http: Action to take if scheme is not
+        ``http`` or ``https`` e.g. ``file:`` or ``about:blank``.
+        If ``True``, the result for non http urls will be an empty string
+        If ``False``, the result for non http urls will be the original url,
+        not further processed e.g. ``about:blank`` -> ``about:blank`` even
+        if ``scheme=False``. The result for http urls will be the stripped
+        url with or without the scheme as per scheme param.
+        Default is ``False``.
+    :type drop_non_http: bool, optional
+    :param use_netloc: If ``True`` urlparse's netloc will be used.
+        If ``False`` urlparse's host will be returned. Using netloc means
+        that a port is included, for example, if it was in the path.
+        Default is ``True``.
+    :type use_netloc: bool, optional
+    :return: Returns a url stripped to (scheme)?+(netloc|hostname)+(path)?.
+        Returns empty string if appropriate.
+    :rtype: str
+    """
     purl = urlparse(url)
-    surl = ''
-    if scheme:
-        surl += purl.scheme + '://'
-    try:
-        surl += purl.hostname + purl.path
-    except TypeError:
-        surl += purl.hostname
-    return surl
+    _scheme = purl.scheme
+
+    # Handle non http schemes
+    if _scheme not in ['http', 'https', '']:
+        if drop_non_http is True:
+            return ''
+        if drop_non_http is False:
+            return url
+
+    if _scheme == '':
+        # From the docs: "urlparse recognizes a netloc only
+        # if it is properly introduced by ‘//’". So we
+        # prepend to get results we expect.
+        url = '//{url}'.format(url=url)
+
+    purl = urlparse(url)
+    scheme_out = ''
+    loc_out = ''
+    path_out = purl.path
+
+    if scheme is True:
+        if _scheme in ['http', 'https']:
+            scheme_out = '{scheme}://'.format(scheme=_scheme)
+        else:
+            # Should only get here if scheme is ''
+            scheme_out = '{scheme}'.format(scheme=_scheme)
```

Comment:
Why still have a format statement here? Can't we just do `scheme_out  = _scheme`?

Reply:
In fact, this `else` condition seems entirely unnecessary since we'll only hit it if `_scheme` is `''` and `scheme_out` defaults to the empty string.